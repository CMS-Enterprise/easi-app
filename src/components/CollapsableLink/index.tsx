import React, { useState } from 'react';
import { Button, Icon } from '@trussworks/react-uswds';
import classNames from 'classnames';

import './index.scss';

type CollapsableLinkProps = {
  id: string;
  className?: string;
  innerClassName?: string;
  children: React.ReactNode | React.ReactNodeArray;
  label: string;
  closeLabel?: string;
  styleLeftBar?: boolean;
  eyeIcon?: boolean;
  startOpen?: boolean;
  labelPosition?: 'top' | 'bottom';
  bold?: boolean;
  onClick?: () => void;
};

const CollapsableLink = ({
  id,
  className,
  innerClassName,
  children,
  label,
  closeLabel,
  styleLeftBar = true,
  eyeIcon,
  startOpen = false,
  labelPosition = 'top',
  bold = true
}: CollapsableLinkProps) => {
  // TODO: should this state instead be held in the parent and passed in as prop?
  // Followup: if the state should remain here, how do we test the component when it's open?
  // That is, how do we initialize this component and set isOpen to true?
  const [isOpen, setOpen] = useState(startOpen);

  const renderEyeIcon = () => {
    return isOpen ? (
      <Icon.VisibilityOff
        className="easi-collapsable-link__eye-icon"
        aria-hidden
      />
    ) : (
      <Icon.Visibility
        className="easi-collapsable-link__eye-icon"
        aria-hidden
      />
    );
  };

  const renderCaret = () => {
    return isOpen ? (
      <Icon.ExpandMore aria-hidden className="margin-right-05" />
    ) : (
      <Icon.NavigateNext className="margin-right-05" aria-hidden />
    );
  };

  const collapseButton: React.ReactNode = (
    <Button
      type="button"
      onClick={() => setOpen(!isOpen)}
      aria-expanded={isOpen}
      aria-controls={id}
      className={classNames(
        'display-flex flex-align-center',
        { 'text-bold': isOpen && bold },
        className
      )}
      unstyled
      data-testid="collapsable-link"
    >
      {eyeIcon ? renderEyeIcon() : renderCaret()}
      {isOpen ? closeLabel || label : label}
    </Button>
  );

  const leftBarStyle = styleLeftBar
    ? 'easi-collapsable-link__content'
    : 'easi-collapsable-link__content-no-bar';

  return (
    <div className="easi-collapsable-link">
      {labelPosition === 'top' && collapseButton}
      {isOpen && (
        <div id={id} className={classNames(innerClassName, leftBarStyle)}>
          {children}
        </div>
      )}
      {labelPosition === 'bottom' && collapseButton}
    </div>
  );
};

export default CollapsableLink;
