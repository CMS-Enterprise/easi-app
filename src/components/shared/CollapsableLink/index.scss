.easi-collapsable-link {
  &__square {
    width: 16px;
    height: 16px;
  }

  &__content {
    border-left: 3px solid #000;
    margin: 4px 0 0 7px;
    padding: 10px 0 10px 6px;
  }
}
