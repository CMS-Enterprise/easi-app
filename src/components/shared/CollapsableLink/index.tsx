import React, { useState } from 'react';
import classnames from 'classnames';
import { Button } from '@trussworks/react-uswds';
import './index.scss';

type CollapsableLinkProps = {
  id: string;
  children: React.ReactNode | React.ReactNodeArray;
  label: string;
};

const CollapsableLink = ({ id, children, label }: CollapsableLinkProps) => {
  // TODO: should this state instead be held in the parent and passed in as prop?
  // Followup: if the state should remain here, how do we test the component when it's open?
  // That is, how do we initialize this component and set isOpen to true?
  const [isOpen, setOpen] = useState(false);
  const arrowClassNames = classnames(
    'fa',
    'easi-collapsable-link__square',
    isOpen ? 'fa-caret-down' : 'fa-caret-right'
  );
  return (
    <div>
      <Button
        type="button"
        onClick={() => setOpen(!isOpen)}
        aria-expanded={isOpen}
        aria-controls={id}
        unstyled
      >
        <span className={arrowClassNames} />
        {label}
      </Button>
      {isOpen && (
        <div id={id} className="easi-collapsable-link__content">
          {children}
        </div>
      )}
    </div>
  );
};

export default CollapsableLink;
