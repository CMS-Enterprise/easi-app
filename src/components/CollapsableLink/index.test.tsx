import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import CollapsableLink from './index';

describe('The Collapsable Link componnet', () => {
  it('renders without crashing', () => {
    const { asFragment } = render(
      <CollapsableLink id="Test" label="testLabel">
        Hello!
      </CollapsableLink>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders children content when expanded', async () => {
    render(
      <CollapsableLink id="Test" label="Test">
        <div data-testid="children" />
      </CollapsableLink>
    );

    userEvent.click(screen.getByTestId('collapsable-link'));
    await screen.findByTestId('children');

    expect(screen.getByRole('button', { name: 'Test' })).toHaveAttribute(
      'aria-expanded',
      'true'
    );
  });
});
