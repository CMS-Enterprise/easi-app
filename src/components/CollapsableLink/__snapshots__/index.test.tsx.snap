// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The Collapsable Link componnet > renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="easi-collapsable-link"
  >
    <button
      aria-controls="Test"
      aria-expanded="false"
      class="usa-button usa-button--unstyled display-flex flex-align-center"
      data-testid="collapsable-link"
      type="button"
    >
      <svg
        class="usa-icon margin-right-05"
        focusable="false"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
        />
      </svg>
      testLabel
    </button>
  </div>
</DocumentFragment>
`;
